packages/core/saved-objects/core-saved-objects-api-server-internal/src/lib/apis/update.ts

```diff
@@ -32,84 +34,146 @@ export interface PerformUpdateParams<T = unknown> {
+    try {
+      response = await executeUpdate(updateParams, apiContext, { namespace });
+      break;
+    } catch (e) {
+      if (
+        SavedObjectsErrorHelpers.isConflictError(e) &&
+        e.retryableConflict &&
+        currentAttempt < maxAttempts
+      ) {
+        continue;
+      }
+      throw e;
+    }
+  }
+
+  return response!;
+};
+
+export const executeUpdate = async <T>(
   { id, type, attributes, options }: PerformUpdateParams<T>,
-  {
-    registry,
-    helpers,
-    allowedTypes,
-    client,
-    serializer,
-    migrator,
-    extensions = {},
-  }: ApiExecutionContext
+  { registry, helpers, client, serializer, extensions = {}, logger }: ApiExecutionContext,
+  { namespace }: { namespace: string | undefined }
 ): Promise<SavedObjectsUpdateResponse<T>> => {
   const {
     common: commonHelper,
     encryption: encryptionHelper,
     preflight: preflightHelper,
     migration: migrationHelper,
+    validation: validationHelper,
   } = helpers;
   const { securityExtension } = extensions;
 
-  const namespace = commonHelper.getCurrentNamespace(options.namespace);
-  if (!allowedTypes.includes(type)) {
-    throw SavedObjectsErrorHelpers.createGenericNotFoundError(type, id);
-  }
-  if (!id) {
-    throw SavedObjectsErrorHelpers.createBadRequestError('id cannot be empty'); // prevent potentially upserting a saved object with an empty ID
-  }
-
   const {
     version,
     references,
     upsert,
     refresh = DEFAULT_REFRESH_SETTING,
-    retryOnConflict = version ? 0 : DEFAULT_RETRY_COUNT,
+    migrationVersionCompatibility,
   } = options;
 
-  let preflightResult: PreflightCheckNamespacesResult | undefined;
-  if (registry.isMultiNamespace(type)) {
-    preflightResult = await preflightHelper.preflightCheckNamespaces({
-      type,
-      id,
-      namespace,
-    });
-  }
+  // Preflight calls to get the doc and check namespaces for multinamespace types.
+  const preflightDocResult = await preflightHelper.preflightGetDocForUpdate({
+    type,
+    id,
+    namespace,
+  });
 
-  const existingNamespaces = preflightResult?.savedObjectNamespaces ?? [];
+  const preflightDocNSResult = await preflightHelper.preflightCheckNamespacesForUpdate({
+    type,
+    id,
+    namespace,
+    preflightDocResult,
+  });
 
+  const existingNamespaces = preflightDocNSResult?.savedObjectNamespaces ?? [];
   const authorizationResult = await securityExtension?.authorizeUpdate({
     namespace,
     object: { type, id, existingNamespaces },
   });
 
-  if (
-    preflightResult?.checkResult === 'found_outside_namespace' ||
-    (!upsert && preflightResult?.checkResult === 'not_found')
-  ) {
+  // validate if an update (directly update or create the object instead) can be done, based on if the doc exists or not
+  // extract into canUpdate and canUpsert method instead where it mustn't be possible for BOTH to be true. It's either an update existing thing or create non-existing thing
+  // can perform request START
```

Comment:
should we remove this comment, the lines above feel more descriptive and useful?